Add missing translations to HashAlgorithmNames

IDisposable opened the pull request:
In PR #84132, the addition of SHA3 variants to the HashAlgorithmNames was incomplete, leaving out the explicit handling of the SHA3 variants from the two helper methods:
  - `ToUpper()` - Would just return the name **unchanged** (which is likely wrong as it would allow _sha3-256_ through instead of uppercasing to _SHA3-256_. That's probably wrong.
- `ToAlgorithmName()` - Would return the `ToString()` on the supplied `hashAlgorithm` instance, which would be the classname of the argument `hashAlgorithm` (e.g. `"System.Security.Cryptography.SHA3_256"`). That would not match the expectation of `"SHA3-256"` based on the other instances.
I am not sure how important these are, or if they should be wrapped in a `#if NET8_0_OR_GREATER` wrapper (only some of the new SHA3 stuff was wrapped in that PR)

adamsitnik commented on src/libraries/System.Security.Cryptography/tests/HashAlgorithmNamesTests.cs:
```diff
@@ -0,0 +1,92 @@
+
+            if (hashSupported)
+            {
+                using (HashAlgorithm hashAlgorithm = CreateHashAlgorithm(hashAlgorithmName);
```
missing `)`

```suggestion
                using (HashAlgorithm hashAlgorithm = CreateHashAlgorithm(hashAlgorithmName));
```

IDisposable replied:
Not sure how this happened, many apologies for the bad rebase...

adamsitnik commented on src/libraries/System.Security.Cryptography/tests/HashAlgorithmNamesTests.cs:
```diff
@@ -0,0 +1,92 @@
+        [Fact]
+        public static void ToAlgorithmName_ToStringUnknownInput()
+        {
+            using (HashAlgorithm hashAlgorithm = CreateHashAlgorithm(UnsupportedHashAlgorithm.ID);
```
```suggestion
            using (HashAlgorithm hashAlgorithm = CreateHashAlgorithm(UnsupportedHashAlgorithm.ID));
```

adamsitnik commented on src/libraries/System.Security.Cryptography/tests/HashAlgorithmNamesTests.cs:
```diff
@@ -0,0 +1,92 @@
+
+        private static HashAlgorithm CreateHashAlgorithm(string hashAlgorithmName)
+        {
+            return hashAlgorithmName switch
+                HashAlgorithmNames.MD5 => MD5.Create(),
+                HashAlgorithmNames.SHA1 => SHA1.Create(),
+                HashAlgorithmNames.SHA256 => SHA256.Create(),
+                HashAlgorithmNames.SHA384 => SHA384.Create(),
+                HashAlgorithmNames.SHA512 => SHA512.Create(),
+                HashAlgorithmNames.SHA3_256 => SHA3_256.Create(),
+                HashAlgorithmNames.SHA3_384 => SHA3_384.Create(),
+                HashAlgorithmNames.SHA3_512 => SHA3_512.Create(),
+                UnsupportedHashAlgorithm.Name => new UnsupportedHashAlgorithm();
```
this should hopefully fix the compilation error

```suggestion
            return hashAlgorithmName switch
            {
                HashAlgorithmNames.MD5 => MD5.Create(),
                HashAlgorithmNames.SHA1 => SHA1.Create(),
                HashAlgorithmNames.SHA256 => SHA256.Create(),
                HashAlgorithmNames.SHA384 => SHA384.Create(),
                HashAlgorithmNames.SHA512 => SHA512.Create(),
                HashAlgorithmNames.SHA3_256 => SHA3_256.Create(),
                HashAlgorithmNames.SHA3_384 => SHA3_384.Create(),
                HashAlgorithmNames.SHA3_512 => SHA3_512.Create(),
                _ => new UnsupportedHashAlgorithm()
            };
```

adamsitnik reviewed (commented):
Overall the change looks good, but we need to fix the build issue before making any progress here.

Thank you for your contribution @IDisposable !

vcsjones reviewed (commented):
From the conversation in the previous pull request at https://github.com/dotnet/runtime/pull/87237, I had asked why we need to make this change.

> I left out modifying ToUpper intentionally because as far as I was able [to observe](https://github.com/search?q=repo%3Adotnet%2Fruntime%20%2FHashAlgorithmNames.ToUpper%2F&type=code) it's only used in RSAPKCS1Signature{De}formatter. These types are only relevant in System.Security.Cryptography.Xml and CryptoConfig. CryptoConfig and S.S.C.Xml isn't getting SHA3 support yet, because Windows didn't add any SHA3 OIDs to their OID registry, and none of the XML specs have been updated to support SHA-3. So they will continue to not support SHA-3 in the same way that .NET Framework does not.

> The same goes for ToAlgorithmName.

> To set the expectation though, there are likely going to be many places you will encounter where we handling SHA1... SHA2... MD5... but it looks like we "forgot" about SHA3. A good number of those places are going to be intentional - Windows hasn't wired SHA3 in to its whole stack yet (like X.509 or EncryptedXml, etc). So we continue to operate as we always have, especially where we want exception-for-exception compatibility with .NET Framework.


HashAlgorithmNames is not a public class, so you cannot test it directly. What actual public API surface needs these changes in order to support SHA-3?

IDisposable commented:
> HashAlgorithmNames is not a public class, so you cannot test it directly. What actual public API surface needs these changes in order to support SHA-3?

I completely forgot I had pushed this up once before, sorry about that. I was cleaning up stale branches and noticed this wasn't merged yet. I am happy to drop this entirely, or follow up with more changes as desired. Sorry about that folks.

IDisposable committed:
Add missing translations to HashAlgorithmNames

In PR #84132, the addition of SHA3 variants to the HashAlgorithmNames was incomplete, leaving out the explicit handling of the SHA3 variants from the two helper methods:
  - `ToUpper()` - Would just return the name **unchanged** (which is likely wrong as it would allow _sha3-256_ through instead of uppercasing to _SHA3-256_. That's probably wrong.
- `ToAlgorithmName()` - Would return the `ToString()` on the supplied `hashAlgorithm` instance, which would be the classname of the argument `hashAlgorithm` (e.g. `"System.Security.Cryptography.SHA3_256"`). That would not match the expectation of `"SHA3-256"` based on the other instances.
I am not sure how important these are, or if they should be wrapped in a `#if NET8_0_OR_GREATER` wrapper (only some of the new SHA3 stuff was wrapped in that PR)

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

vcsjones commented:
>  I am happy to drop this entirely, or follow up with more changes as desired. Sorry about that folks.

No worries, and as I said in the other pull request, I'm always happy when people take an interest in this area.

For now though, I think we can close this pull request. We can re-open it when and if it can be demonstrated these changes are necessary to enable SHA-3 in an area of the API surface where it is desirable and possible to fully support SHA-3. Testing `HashAlgorithmNames` is not to fulfill this - it's just a utility class used in an area of the code (S.S.C.Xml and CryptoConfig) where SHA-3 is not being applied.
